xdp-loader/xdp-loader.c

```diff
@@ -219,8 +218,26 @@ int do_unload(const void *cfg, __unused const char *pin_root_path)
 		struct xdp_program *prog = NULL;
 
 		while ((prog = xdp_multiprog__next_prog(prog, mp))) {
-			if (xdp_program__id(prog) == opt->prog_id)
+			if (xdp_program__id(prog) == opt->prog_id) {
+				mode = xdp_multiprog__attach_mode(mp);
 				break;
+			}
+		}
+		if (!prog && xdp_multiprog__is_legacy(mp)) {
+			prog = xdp_multiprog__main_prog(mp);
+			if (xdp_program__id(prog) == opt->prog_id) {
+				mode = xdp_multiprog__attach_mode(mp);
+			} else {
+				prog = NULL;
+			}
+		}
+		if (!prog) {
+			prog = xdp_multiprog__hw_prog(mp);
+			if (xdp_program__id(prog) == opt->prog_id) {
+				mode = XDP_MODE_HW;
+			} else {
+				prog = NULL;
+			}
 		}
```

Comment:
Instead of setting and reset'ing 'prog' I think you can simplify this a bit by just using 'goto found' if you find the right program...